### ISSUE TITLE
Marvin AI et Outlines — sortie structurée fiable

### ISSUE THREAD

**Issue by khalilbenaz** [khalilbenaz/khalilbenaz.github.io#170]
---
lede: Marvin AI (haute-ergonomie) et Outlines (state-machine) forcent les LLMs à produire des sorties valides selon un schéma. Indispensable en prod.
lede_en: Marvin AI (high-ergonomics) and Outlines (state-machine) force LLMs to produce schema-valid outputs. Essential in prod.
title_en: Marvin AI and Outlines — reliable structured output
---

## Le problème de la sortie LLM

Vous voulez du JSON depuis un LLM. Sans contraintes, le modèle peut :

- Oublier une clé.
- Renvoyer une string au lieu d'un nombre.
- Mettre du Markdown ` ``` ` autour.
- Hallucinser un champ inexistant.

Solutions classiques : retry sur validation Pydantic/Zod. Ça marche mais coûte des tokens et de la latence.

Solutions avancées : forcer le modèle à respecter un schéma au niveau **token**.

## Marvin AI — ergonomie maximale

Marvin (Python) propose des décorateurs naturels :

```python
import marvin
from pydantic import BaseModel

class Person(BaseModel):
    name: str
    age: int
    city: str

@marvin.fn
def extract_person(text: str) -> Person:
    '''Extract person from text.'''

result = extract_person("Khalil, 32 ans, vit à Rabat.")
# Person(name='Khalil', age=32, city='Rabat')
```

Marvin :

- Compose le prompt automatiquement à partir de la docstring et du return type.
- Appelle le LLM (Anthropic, OpenAI, etc.).
- Parse, valide via Pydantic.
- Retry intelligent si rate.

DX excellente pour des extractions courantes.

## Outlines — la contrainte au niveau token

Outlines (Python) va plus loin techniquement : il **contraint la génération token par token** selon une grammaire ou un schéma.

```python
import outlines

model = outlines.models.transformers("meta-llama/Llama-3.1-8B")
generator = outlines.generate.json(model, Person)
result = generator("Khalil, 32 ans, vit à Rabat.")
# Strictement Person, garanti par construction
```

Sous le capot : à chaque token généré, Outlines masque les tokens qui briseraient le JSON schema. Le modèle ne peut pas dévier.

Conséquences :

- **100 % valide par construction** : pas de retry nécessaire.
- **Plus rapide** : pas de tokens "Markdown" ou "explication" générés.
- **Schémas complexes** : enums, regex, JSON Schema, BNF grammars supportés.

Limite : ne marche qu'avec des modèles dont vous avez accès aux logits (local LLM, vLLM, certains providers exposant logit_bias). Pas avec Claude ou GPT-4 directement.

## Quand utiliser quoi

- **Vous appelez Claude/OpenAI/Gemini en cloud** : Marvin AI ou Instructor. Plus simple, fonctionne partout.
- **Vous tournez du LLM local** (Llama, Mistral) : Outlines pour garantie 100 %.
- **Production high-volume** : Outlines + local LLM = pas de retry = économie.

## Vs alternatives

- **Pydantic AI** : très proche de Marvin, plus minimaliste.
- **Instructor** : équivalent côté TypeScript et Python, ergonomie proche de Marvin.
- **OpenAI Structured Outputs** : natif si vous êtes OpenAI-only, supporte JSON Schema strict.
- **Anthropic Tool Use** : trick — déclarer un tool qui prend votre struct en argument.

Pour un projet cross-provider avec budget tokens : Marvin / Instructor. Pour un projet local-LLM ou ultra-volume : Outlines.

## The LLM output problem

You want JSON from an LLM. Without constraints, the model can:

- Forget a key.
- Return a string instead of a number.
- Wrap output in Markdown ` ``` `.
- Hallucinate a non-existent field.

Classic solutions: retry on Pydantic/Zod validation. Works but costs tokens and latency.

Advanced solutions: force the model to respect a schema at the **token** level.

## Marvin AI — maximum ergonomics

Marvin (Python) offers natural decorators:

```python
import marvin
from pydantic import BaseModel

class Person(BaseModel):
    name: str
    age: int
    city: str

@marvin.fn
def extract_person(text: str) -> Person:
    '''Extract a person from text.'''

result = extract_person("Khalil, 32, lives in Rabat.")
# Person(name='Khalil', age=32, city='Rabat')
```

Marvin:

- Composes the prompt automatically from docstring and return type.
- Calls the LLM (Anthropic, OpenAI, etc.).
- Parses, validates via Pydantic.
- Smart retry on failure.

Excellent DX for common extractions.

## Outlines — token-level constraint

Outlines (Python) goes further technically: it **constrains generation token-by-token** per a grammar or schema.

```python
import outlines

model = outlines.models.transformers("meta-llama/Llama-3.1-8B")
generator = outlines.generate.json(model, Person)
result = generator("Khalil, 32, lives in Rabat.")
# Strictly Person, guaranteed by construction
```

Underneath: for each generated token, Outlines masks tokens that would break the JSON schema. The model can't deviate.

Consequences:

- **100% valid by construction**: no retry needed.
- **Faster**: no "Markdown" or "explanation" tokens generated.
- **Complex schemas**: enums, regex, JSON Schema, BNF grammars supported.

Limit: only works with models you have logit access to (local LLM, vLLM, some providers exposing logit_bias). Not directly with Claude or GPT-4.

## When to use what

- **Calling cloud Claude/OpenAI/Gemini**: Marvin AI or Instructor. Simpler, works everywhere.
- **Running local LLM** (Llama, Mistral): Outlines for 100% guarantee.
- **High-volume production**: Outlines + local LLM = no retry = savings.

## vs alternatives

- **Pydantic AI**: very close to Marvin, more minimalist.
- **Instructor**: equivalent on TypeScript and Python side, ergonomics close to Marvin.
- **OpenAI Structured Outputs**: native if OpenAI-only, supports strict JSON Schema.
- **Anthropic Tool Use**: trick — declare a tool taking your struct as argument.

Cross-provider project with token budget: Marvin / Instructor. Local-LLM or ultra-volume project: Outlines.
